### DISCUSSION TITLE
how to debug c++ code in milvus

### DISCUSSION THREAD

**Discussion by cherrylzhao** [milvus-io/milvus#38660] [Q&A and General discussion] [answered]
hi, everyone
I have modified the mode to "Debug" in Makefile, then compiled milvus 2.4.18 in centos 7.9 successfully.
currently, Attu can connect to my local compiled milvus binary.
so I want to investigate the vector query process internal, I try to attach the pid of milvus in gdb,
but it seems that breakpoint in c++ code could not stop, please give me some hint of how can I correctly debug c++ code.

thanks

**Comment by yhmo** [comment]
Milvus process is running with a lot of parallel threads, if you set a breakpoint to a thread, the other thread might encounter timeout error and run into a weird state. So, milvus developers don't "attach and debug", they add temp logs in source code and print the runtime variables to diagnose, or review the full log to investigate.

**Reply by cherrylzhao** [reply]
thanks to reply !
currently, I just want to know the invoke chain of query process in source code, could you please give me some sample debug logs of that?

**Reply by yhmo** [reply]
All requests are received by proxy node, the search request entry is here:
https://github.com/milvus-io/milvus/blob/bb5f38e57433341bbadeccb0b15c2185944d6744/internal/proxy/impl.go#L2991

Once a search request is received, a searchTask is put into a scheduler to queue:
https://github.com/milvus-io/milvus/blob/master/internal/proxy/task_scheduler.go
https://github.com/milvus-io/milvus/blob/master/internal/proxy/task_search.go

searchTask.searchShard() is called if the searchTask is to be executed:
https://github.com/milvus-io/milvus/blob/bb5f38e57433341bbadeccb0b15c2185944d6744/internal/proxy/task_search.go#L767

A query node's Search() is called by searchTask.searchShard():
https://github.com/milvus-io/milvus/blob/bb5f38e57433341bbadeccb0b15c2185944d6744/internal/querynodev2/services.go#L740

ShardDelegator.Serch() is called. Each shard delegator is a leader node for a shard of a collection. By default, a collection's shard_number is 1, there is only one shard delegator. If shard_number is N, there will be N shard delegators work for the search request.
https://github.com/milvus-io/milvus/blob/82d85e8c57b8c144e53789fe725960d5c0600a86/internal/querynodev2/delegator/delegator.go#L273

remoteWorker.SearchSegments() is called by the shardDelegator.search()
https://github.com/milvus-io/milvus/blob/82d85e8c57b8c144e53789fe725960d5c0600a86/internal/querynodev2/cluster/worker.go#L191

QueryNode.SearchSegments() is called by emoteWorker.SearchSegments() 
https://github.com/milvus-io/milvus/blob/82d85e8c57b8c144e53789fe725960d5c0600a86/internal/querynodev2/services.go#L655

SearchTask is created by QueryNode.SearchSegments() and executed
https://github.com/milvus-io/milvus/blob/82d85e8c57b8c144e53789fe725960d5c0600a86/internal/querynodev2/tasks/search_task.go#L133

Segment.search() is called by searchSegments()
https://github.com/milvus-io/milvus/blob/82d85e8c57b8c144e53789fe725960d5c0600a86/internal/querynodev2/segments/search.go#L48

The implementation of Segment.search(), from this method, you will see the code runs into CGo
https://github.com/milvus-io/milvus/blob/82d85e8c57b8c144e53789fe725960d5c0600a86/internal/querynodev2/segments/segment.go#L492C10-L492C22

The search entry in CGo:
https://github.com/milvus-io/milvus/blob/82d85e8c57b8c144e53789fe725960d5c0600a86/internal/core/src/segcore/segment_c.cpp#L110

**Reply by cherrylzhao** [reply, marked as answer]
thanks for your detail explain, that help me a lot !
